### PR TITLE
fix ip length and static cast bugs, add processor arg for message allback

### PR DESCRIPTION
**Pull request by zjusuyong**
fix ip length and static cast bugs, add processor arg for message callback

### CODE REVIEW

File: reader.c [hunk 1]
```diff
@@ -86,20 +87,21 @@ static void nsq_reader_lookupd_poll_cb(EV_P_ struct ev_timer *w, int revents)
     ev_timer_again(rdr->loop, &rdr->lookupd_poll_timer);
 }
 
-struct NSQReader *new_nsq_reader(struct ev_loop *loop, const char *topic, const char *channel,
+struct NSQReader *new_nsq_reader(struct ev_loop *loop, const char *topic, const char *channel, void *processor,
     void (*connect_callback)(struct NSQReader *rdr, struct NSQDConnection *conn),
     void (*close_callback)(struct NSQReader *rdr, struct NSQDConnection *conn),
-    void (*msg_callback)(struct NSQReader *rdr, struct NSQDConnection *conn, struct NSQMessage *msg))
+    void (*msg_callback)(struct NSQReader *rdr, struct NSQDConnection *conn, struct NSQMessage *msg, void *processor))
```
Comment (mreiferson) [inline reader.c]: I don't understand what `processor` is?

**Comment by mreiferson** [comment]
Thanks for this!

File: reader.c (hunk 1 above)
Comment (zjusuyong) [inline reader.c]: that can be a function pointer or c++ object pointer, in order to process the message.

Comment (mreiferson) [inline reader.c]: I'm not sure I follow - that's what `msg_callback` is already?

Comment (mreiferson) [inline reader.c]: Can you provide an example of how you expect to use `processor`?

Comment (zjusuyong) [inline reader.c]: //handle message
static void message_handler(struct NSQReader *rdr, struct NSQDConnection *conn, struct NSQMessage
        *msg, void *msg_processor){
    NsqProcessor * nsqprocessor = (NsqProcessor *)msg_processor;
    int ret = nsqprocessor->process(msg);
    buffer_reset(conn->command_buf);
    if(ret < 0){
        //重新放回，超时为100ms
        LOG4CXX_ERROR(log4cxx::Logger::getRootLogger(),
            "consumer message_process failed, call nsq_requeue for msg_id:" << msg->id);
        nsq_requeue(conn->command_buf, msg->id, 100);
    }else{
        LOG4CXX_DEBUG(log4cxx::Logger::getRootLogger(),
            "consumer message_process succeed, call nsq_finish for msg_id:" << msg->id);
        nsq_finish(conn->command_buf, msg->id);
    }

    buffered_socket_write_buffer(conn->bs, conn->command_buf);
    buffer_reset(conn->command_buf);
    nsq_ready(conn->command_buf, rdr->max_in_flight);
    buffered_socket_write_buffer(conn->bs, conn->command_buf);
    free_nsq_message(msg);
}

//to start consumer
void start_consume(const std::string & topic, const std::string & channel, const std::string &ip, int port){
       struct ev_loop *loop = ev_loop_new(0);
       struct NSQReader * rdr = new_nsq_reader(loop, topic.c_str(), channel.c_str(), msg_processor,
        NULL, NULL, message_handler);
      nsq_reader_add_nsqlookupd_endpoint(rdr, ip.c_str(), port);
}

class NsqProcessor{
public:
    NsqProcessor();
    ~NsqProcessor();
     int process(void *msg, size_t msg_length=0);

private:
    int count;  //for example
};

Comment (mreiferson) [inline reader.c]: Oh, OK, it's basically "context".

Can we rename the variable to `void *ctx`?

Comment (zjusuyong) [inline reader.c]: ok

**Pull request merged**